pkg/model/deployment.go

```diff
@@ -0,0 +1,207 @@
+//
+// Copyright (c) 2023 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+	"os"
+
+	corev1 "k8s.io/api/core/v1"
+
+	bsv1alpha1 "redhat-developer/red-hat-developer-hub-operator/api/v1alpha1"
+
+	"redhat-developer/red-hat-developer-hub-operator/pkg/utils"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const BackstageImageEnvVar = "RELATED_IMAGE_backstage"
+const defaultMountDir = "/opt/app-root/src"
+
+type BackstageDeploymentFactory struct{}
+
+func (f BackstageDeploymentFactory) newBackstageObject() RuntimeObject {
+	return &BackstageDeployment{}
+}
+
+type BackstageDeployment struct {
+	deployment *appsv1.Deployment
+}
+
+func init() {
+	registerConfig("deployment.yaml", BackstageDeploymentFactory{})
+}
+
+func DeploymentName(backstageName string) string {
+	return utils.GenerateRuntimeObjectName(backstageName, "backstage")
+}
+
+// implementation of RuntimeObject interface
+func (b *BackstageDeployment) Object() client.Object {
+	return b.deployment
+}
+
+func (b *BackstageDeployment) setObject(obj client.Object) {
+	b.deployment = nil
+	if obj != nil {
+		b.deployment = obj.(*appsv1.Deployment)
+	}
+}
+
+// implementation of RuntimeObject interface
+func (b *BackstageDeployment) EmptyObject() client.Object {
+	return &appsv1.Deployment{}
+}
+
+// implementation of RuntimeObject interface
+func (b *BackstageDeployment) addToModel(model *BackstageModel, _ bsv1alpha1.Backstage) (bool, error) {
+	if b.deployment == nil {
+		return false, fmt.Errorf("Backstage Deployment is not initialized, make sure there is deployment.yaml in default or raw configuration")
+	}
+	model.backstageDeployment = b
+	model.setRuntimeObject(b)
+
+	// override image with env var
+	// [GA] TODO Do we need this feature?
+	if os.Getenv(BackstageImageEnvVar) != "" {
+		b.deployment.Spec.Template.Spec.Containers[0].Image = os.Getenv(BackstageImageEnvVar)
+		// exactly the same image for initContainer and want it to be overriden
+		// the same way as Backstage's one
+		for i := range b.deployment.Spec.Template.Spec.InitContainers {
+			b.deployment.Spec.Template.Spec.InitContainers[i].Image = os.Getenv(BackstageImageEnvVar)
+		}
+	}
+
+	return true, nil
+}
+
+// implementation of RuntimeObject interface
+func (b *BackstageDeployment) validate(model *BackstageModel, backstage bsv1alpha1.Backstage) error {
```

Comment:
Why is this method called `validate`, while it actually mutates `b`? I would expect it to just, well, validate things and return errors.. :)

Reply:
You are right, the name is not perfect.
I am still thinking about this name. Something like "mutate" is not correct either because it is a second part of mutate.
Let's keep it as it is for initial commit, it is not a part of public API so can be easily changed any moment.
Proposals welcomed ;)